# Урок 2. Keras

## Практическое задание

<ol>
    <li>Попробуйте обучить, нейронную сеть на Keras(рассмотренную на уроке) на датасете MNIST с другими параметрами. 
        Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность?</li>
    <li>Поработайте с документацией Keras. Попробуйте найти полезные команды Keras неразобранные на уроке.</li>
</ol>

In [67]:
import numpy as np
import mnist

from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.callbacks import EarlyStopping
from keras import backend

In [68]:
# Load
train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

In [69]:
# Build the model.
model = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax'),
])

model.summary()

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_29 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_30 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_31 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [70]:
# Compile the model.
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy'],
)

In [71]:
# После двух эпох без улучшения точности в 0.001 обучение прекратится
early_stopping_monitor = EarlyStopping(monitor='accuracy', patience=2, min_delta=0.001)

# Train the model.
model.fit(
    train_images,
    to_categorical(train_labels),
    epochs=30,
    batch_size=32,
    callbacks=[early_stopping_monitor]
)

# Save the model to disk.
model.save_weights('model.h5')

Epoch 1/30
60000/60000 [==============================] - 4s 64us/step - loss: 0.3219 - accuracy: 0.9027
Epoch 2/30
60000/60000 [==============================] - 4s 67us/step - loss: 0.1625 - accuracy: 0.9503
Epoch 3/30
60000/60000 [==============================] - 4s 65us/step - loss: 0.1243 - accuracy: 0.9616
Epoch 4/30
60000/60000 [==============================] - 4s 62us/step - loss: 0.1008 - accuracy: 0.9688
Epoch 5/30
60000/60000 [==============================] - 4s 69us/step - loss: 0.0851 - accuracy: 0.9733
Epoch 6/30
60000/60000 [==============================] - 4s 64us/step - loss: 0.0774 - accuracy: 0.9747
Epoch 7/30
60000/60000 [==============================] - 4s 60us/step - loss: 0.0686 - accuracy: 0.9779
Epoch 8/30
60000/60000 [==============================] - 4s 62us/step - loss: 0.0628 - accuracy: 0.9797
Epoch 9/30
60000/60000 [==============================] - 4s 66us/step - loss: 0.0578 - accuracy: 0.9803
Epoch 10/30
60000/60000 [==============================

In [72]:
# Evaluate the model.
test_loss, test_acc = model.evaluate(
    test_images,
    to_categorical(test_labels)
)

print('Ошибка на тестовых данных:', test_loss)
print('Точность на тестовых данных:', test_acc)

10000/10000 [==============================] - 0s 41us/step
Ошибка на тестовых данных: 0.08545002939769201
Точность на тестовых данных: 0.9786999821662903


#### Результат который получается при исходном обучении

#### Train the model.
model.fit(<br>
    train_images,<br>
    to_categorical(train_labels),<br>
    epochs=5,<br>
    batch_size=32<br>
)<br>

#### Save the model to disk.
model.save_weights('model.h5')<br>

Epoch 1/5<br>
60000/60000 [==============================] - 6s 93us/step - loss: 0.3532 - accuracy: 0.8938<br>
Epoch 2/5<br>
60000/60000 [==============================] - 5s 80us/step - loss: 0.1859 - accuracy: 0.9438<br>
Epoch 3/5<br>
60000/60000 [==============================] - 3s 57us/step - loss: 0.1432 - accuracy: 0.9561<br>
Epoch 4/5<br>
60000/60000 [==============================] - 4s 74us/step - loss: 0.1191 - accuracy: 0.9633 0s - loss: 0.1<br>
Epoch 5/5<br>
60000/60000 [==============================] - 4s 63us/step - loss: 0.1034 - accuracy: 0.9680

#### Evaluate the model.
test_loss, test_acc = model.evaluate(<br>
    test_images,<br>
    to_categorical(test_labels)<br>
)<br>

print('Ошибка на тестовых данных:', test_loss)<br>
print('Точность на тестовых данных:', test_acc)<br>

10000/10000 [==============================] - 0s 32us/step<br>
Ошибка на тестовых данных: 0.10802537681944668<br>
Точность на тестовых данных: 0.9661999940872192

##### Улучшил результат с 0.9661999940872192 до 0.9786999821662903

In [74]:
# Load the model from disk later using:
# model.load_weights('model.h5')

# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


In [75]:
# predict_classes позволяет срузу получить метки классов
predictions = model.predict_classes(test_images[:5])
predictions

array([7, 2, 1, 0, 4], dtype=int64)

In [76]:
# Посмотрим на доступную видео-карту
backend.tensorflow_backend._get_available_gpus()

['/job:localhost/replica:0/task:0/device:GPU:0']

## Результаты практического задания:
1. Увеличил точность модели на тестовых данных с 0.9664999842643738, до 0.9786999821662903. Этого удалось добиться за счёт увеличения чила эпох и числа нейронов в первом слое.<br>
2. Полезные команды Keras неразобранные на уроке:<br>
    2.1. Из метода evaluate получены параметры для оценки ошибки и точности тестовых данных.<br>
    2.2. EarlyStopping позволяет закончить обучение если результаты больше не улучшаются.<br>
    2.3. predict_classes позволяет срузу получить метки классов<br>
    2.4. backend.tensorflow_backend._get_available_gpus позволяет увидеть доступные видеокарты